[Ready] Adds a new power, Transcendent Olfaction, to genetics

PKPenguin321 opened the pull request:
codersprite by me

:cl: PKPenguin321
add: Hunt down your enemies by scent with a new power that has been added to genetics: Transcendent Olfaction. 
/:cl:

You know how dogs have a really strong sense of smell, and how in the movies people will hold a handkerchief up to them or something and then they smell it and then suddenly go running after whoever owned the handkerchief despite them being dozens of miles away? This is kind of like that.

Pick up an item that was held by somebody else and has their fingerprints on it. Then you can catch their scent, and subsequent uses will tell you what direction they are from you in terms of north, south, east, west, etc. Each use has a 10 second cooldown, and I had a lot of fun writing the dialogue.

The big downside is that hygiene is more punishing, due to your nose being way more sensitive. If you start getting stink lines, you'll rapidly become very disgusted, which can (in very extreme cases) cause frequent vomiting.

PKPenguin321 committed:
adds Transcendent Olfaction to genetics
codersprite by me

imsxz commented:
snnnnnnnnnniiiiiiiiiiiiiiiiiiiiiiiiiiiiiiifff

GuyonBroadway commented:
Lets get the **"FETISH CONTENT"** joke out the way.

GuyonBroadway commented:
Also it would be funny if you could get Ian to sniff something and he would try to path to thier location.

subject217 commented:
BBBBBBBBBRRRRRRRRRRRAAAAAAAAAAAPPPPPPPPPPPPPPPPPsnnnnniiiiiiffffffffffff...oh yes my dear....sssnnnnnnnnnnnniiiiiiiiffffffff....quite pungent indeed...is that....dare I say....sssssssnniff...eggs I smell?......sniff sniff....hmmm...yes...quite so my darling....sniff....quite pungent eggs yes very much so .....ssssssssssssssnnnnnnnnnnnnnnniiiiiiiffffff....ah yes...and also....a hint of....sniff....cheese.....quite wet my dear....sniff...but of yes...this will do nicely....sniff.....please my dear....another if you please....nice a big now....BBBBBBRRRRRRRAAAAAAAPPPPPPPFFFFFFFFLLLLLLLLLPPPPPPPPPFFFFFF Oh yes...very good!....very sloppy and wet my dear....hmmmmm...is that a drop of nugget I see on the rim?...hmmmm.....let me.....let me just have a little taste before the sniff my darling.......hmmmmm....hmm..yes....that is a delicate bit of chocolate my dear....ah yes....let me guess...curry for dinner?....oh quite right I am....aren't I?....ok....time for sniff.....sssssnnnnnnniiiiiiiiffffffff.....hmmm...hhhmmmmm I see...yes....yes indeed as well curry......hmmm....that fragrance is quite noticeable....yes.....onion and garlic chutney I take it my dear?.....hmmmmm....yes quite.....BBBBBBRRRRRRRRPPPPPPFFFFFFFFFFFFFFFFFFFFFTTTTTTTTTTT Oh I was not expecting that…that little gust my dear….you caught me off guard…yes…so gentle it was though…hmmmm…let me taste this little one…just one small sniff…..sniff…ah….ssssssnnnnnniiiiiffffffffffff…and yet…so strong…yes…the odor….sniff sniff…hmmm….is that….sniff….hmmm….I can almost taste it my dear…..yes….just…sniff….a little whiff more if you please…..ssssssnnnnnniiiiiffffffffff…ah yes I have it now….yes quite….hhhhmmmm…delectable my dear…..quite exquisite yes…..I dare say…sniff….the most pungent one yet my dear….ssssnnnnniiiifffffffffffffffffffffff….yes….﻿

PKPenguin321 commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,107 @@
+	if(hygiene_now < HYGIENE_LEVEL_DIRTY && reek >= HYGIENE_LEVEL_DIRTY)
+		to_chat(usr,"<span class='warning'>Your horrible stench causes your nostrils to slam shut as your survival instincts involuntarily kick in.</span>")
+		to_chat(usr,"<span class='userdanger'>You can't breathe!</span>")
+		reek = hygiene_now
```
OK so basically I'm dumb, just need to move this back a tab and delete all the other instances of it in this whole block

PKPenguin321 replied:
Also I accidentally added a duplicate sprite

GuyonBroadway commented:
Rather than make poor hygiene kill you why not make it gross you out so you puke with debilitating regularity?

Shadowflame909 commented:
The detective would love this power

HomocidlePancakes commented:
Wew. Buff to wolfgirls imho

OverlyGeneric commented:
This is taking "I can smell you" to a creepier level for traitors now.

Shardj commented:
Give Ian this power so you can pop him in the genetics machine to find it

PKPenguin321 commented:
Sadly I don't think Ian is compatible with the genetics system so that would either not work or be pretty snowflakey

PKPenguin321 committed:
fixes
made messages work, added a new message
removed duplicate icon

necromanceranne commented:
*I can smell your fear...*

Wilchenx commented:
>Hunt down your enemies by scent

fetish content?

PKPenguin321 committed:
uses disgust instead of oxy damage

PKPenguin321 commented:
The latest commit switches the penalty from outright killing you with oxygen damage to making you disgusted.

Wilchenx commented:
@GuyonBroadway whoosh

4dplanner commented:
Off topic, but #define VOID		/datum/mutation/human/void as opposed to MUTATION_VOID is gross

Qustinnus commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+	if(hygiene_now < 100 && prob(5))
+		owner.adjust_disgust(rand(3,5))
+	if(hygiene_now < HYGIENE_LEVEL_DIRTY && prob(50))
+		to_chat(owner,"<span class='danger'>You get a whiff of your stench and feel sick!</span>")
```
wouldnt this happen once every 4 seconds on average?

PKPenguin321 replied:
Yeah, but you get a shitload of warning in advance and even then it takes a good amount of disgust before you start vomiting. I tested it by setting myself to 70 hygiene and I was able to cross the station and make it to a shower with about 120~ disgust total, which is a fair penalty for not washing yourself for literal dozens of minutes despite multiple warnings..

Qustinnus replied:
i ment the to_chats, also dw I made disgust I know how long it takes before you puke :^)

PKPenguin321 replied:
The repeated messages concatenate anyways and drive how urgent it is (since if you ignore it you WILL get pretty much chainstunned by vomit for a long time)

datorangebottle commented:
I like the idea of tracking people via scent but I dislike the idea of tying in hygiene- I honestly don't think you should be coding to involve a feature that a very vocal portion of the playerbase mocks and hates on a regular basis, to the point that it's a meme on Sybil even after the nerf.

Qustinnus commented:
"this feature is disliked so it shouldnt get more interesting content"

nice.

Jordie0608 commented:
I smell some conflicts here.

PKPenguin321 commented:
Yep. Will have to update it now that one of the three PRs I had up was merged. Expect it to be resolved in maybe 7-8 hours

PKPenguin321 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into sniffff-but-not-a-brap-meme

# Conflicts:
#	code/__DEFINES/DNA.dm
#	code/game/objects/items/dna_injector.dm
#	icons/mob/actions/actions_spells.dmi

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+	if(hygiene_now < 100 && reek >= 100)
+		to_chat(owner,"<span class='danger'>Your odor begins to make you gag. You silently curse your godly nose. You should really get clean!</span>")
+	if(hygiene_now < HYGIENE_LEVEL_DIRTY && reek >= HYGIENE_LEVEL_DIRTY)
+		to_chat(owner,"<span class='danger'>Your super-nose is fed up with your stench. You could really use a shower.</span>")
```
Shouldn't these be else ifs? It seems you'll be getting spammed all 4 at the highest level

PKPenguin321 replied:
No. The `reek` var tracks your hygiene amount from the previous tick so that these messages are only shown once when you pass a threshold. If you somehow pass two thresholds in a tick, I'd like multiple messages to show.

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+	var/mob/living/carbon/tracking_target
+	var/list/mob/living/carbon/possible = list()
+
+/obj/effect/proc_holder/spell/targeted/olfaction/cast(list/targets,mob/user = usr)
```
```suggestion
/obj/effect/proc_holder/spell/targeted/olfaction/cast(list/targets, mob/living/user = usr)
```

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+	var/list/mob/living/carbon/possible = list()
+
+/obj/effect/proc_holder/spell/targeted/olfaction/cast(list/targets,mob/user = usr)
+	var/atom/sniffed = usr.get_active_held_item()
```
```suggestion
	var/atom/sniffed = user.get_active_held_item()
```

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+		var/old_target = tracking_target
+		possible = list()
+		var/list/prints = sniffed.return_fingerprints()
+		for(var/mob/living/carbon/C in GLOB.mob_list)
```
```suggestion
		for(var/mob/living/carbon/C in GLOB.carbon_list)
```

or something like that

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+			if(prints[md5(C.dna.uni_identity)])
+				possible |= C
+		if(!length(possible))
+			to_chat(usr,"<span class='warning'>Despite your best efforts, there are no scents to be found on [sniffed]...</span>")
```
```suggestion
			to_chat(user,"<span class='warning'>Despite your best efforts, there are no scents to be found on [sniffed]...</span>")
```

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+		tracking_target = input(user, "Choose a scent to remember.", "Scent Tracking") as null|anything in possible
+		if(!tracking_target)
+			if(!old_target)
+				to_chat(usr,"<span class='warning'>You decide against remembering any scents. Instead, you notice your own nose in your peripheral vision. This goes on to remind you of that one time you started breathing manually and couldn't stop. What an awful day that was.</span>")
```
```suggestion
				to_chat(user,"<span class='warning'>You decide against remembering any scents. Instead, you notice your own nose in your peripheral vision. This goes on to remind you of that one time you started breathing manually and couldn't stop. What an awful day that was.</span>")
```

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+			tracking_target = old_target
+			on_the_trail()
+			return
+		to_chat(usr,"<span class='notice'>You pick up the scent of [tracking_target]. The hunt begins.</span>")
```
```suggestion
		to_chat(user,"<span class='notice'>You pick up the scent of [tracking_target]. The hunt begins.</span>")
```

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+		return
+
+	if(!tracking_target)
+		to_chat(usr,"<span class='warning'>You're not holding anything to smell, and you haven't smelled anything you can track. You smell your palm instead; it's kinda salty.</span>")
```
```suggestion
		to_chat(user,"<span class='warning'>You're not holding anything to smell, and you haven't smelled anything you can track. You smell your palm instead; it's kinda salty.</span>")
```

Cyberboss commented on code/datums/mutations/actions.dm:
```diff
@@ -8,6 +8,95 @@
+
+/obj/effect/proc_holder/spell/targeted/olfaction/proc/on_the_trail()
+	if(!tracking_target)
+		to_chat(usr,"<span class='warning'>You're not tracking a scent, but the game thought you were. Something's gone wrong! Report this as a bug.</span>")
```
Should pass a `user` var and not use usr

PKPenguin321 commented:
Will add these suggestions when I get back to my computer today.

PKPenguin321 committed:
Merge branch 'master' of https://github.com/tgstation/tgstation into sniffff-but-not-a-brap-meme

# Conflicts:
#	icons/mob/actions/actions_spells.dmi

PKPenguin321 committed:
addresses review
mostly changing usr to user. global mob list -> global carbon list since that's what i parse for anyways.
also changed a message.

PKPenguin321 commented:
@Cyberboss Review addressed

PKPenguin321 committed:
makes it compile
